pkg/ddl/rollingback.go

```diff
@@ -339,18 +339,184 @@ func convertAddTablePartitionJob2RollbackJob(d *ddlCtx, t *meta.Meta, job *model
+			// We cannot drop the index here, we need to wait until
+			// the next schema version
+			// i.e. rollback in onDropTablePartition
+			// New index that became public in this state,
+			// mark it to be dropped in next schema version
+			if indexInfo.Global {
+				indexInfo.State = model.StateDeleteReorganization
+			} else {
+				// How to know if this index was created as a duplicate or not?
+				if id, ok := globalToUniqueDupMap[indexInfo.Name.L]; ok {
+					// The original index
+					if id >= indexInfo.ID {
+						return ver, errors.NewNoStackErrorf("Indexes in wrong order during rollback, '%s', %d >= %d", indexInfo.Name.O, id, indexInfo.ID)
+					}
+					indexInfo.State = model.StateDeleteReorganization
+				} else {
+					globalToUniqueDupMap[indexInfo.Name.L] = indexInfo.ID
+				}
+			}
+		}
+	}
+	for _, indexInfo := range dropIndices {
+		DropIndexColumnFlag(tblInfo, indexInfo)
+		RemoveDependentHiddenColumns(tblInfo, indexInfo)
+		removeIndexInfo(tblInfo, indexInfo)
+	}
+	if pi.DDLState == model.StateDeleteReorganization {
+		// New partitions are public,
+		// but old is still double written.
+		// OK to revert.
+		// Remove the AddingDefinitions
+		// Add back the DroppingDefinitions
+		if job.Type == model.ActionReorganizePartition {
+			// Reassemble the list of partitions in the same order:
+			// If RANGE, then there should be the same interval
+			// as was replaced!
+			// If KEY/HASH all are replaced
+			// if LIST it needs the following special handling.
+			if pi.Type == model.PartitionTypeList {
+				// Special handling, since the original partitions might not be in order
+				// they should be in the same order within the two arrays.
+				defPos := 0
+				dropPos := 0
+				newDefs := make([]model.PartitionDefinition, 0, len(pi.OriginalPartitionIDsOrder))
+				for _, id := range pi.OriginalPartitionIDsOrder {
+					if defPos < len(pi.Definitions) && pi.Definitions[defPos].ID == id {
+						newDefs = append(newDefs, pi.Definitions[defPos])
+						defPos++
+						continue
+					}
+					if dropPos < len(pi.DroppingDefinitions) && id == pi.DroppingDefinitions[dropPos].ID {
+						newDefs = append(newDefs, pi.DroppingDefinitions[dropPos])
+						dropPos++
+						continue
+					}
+					for {
+						defPos++
+						if defPos < len(pi.Definitions) && pi.Definitions[defPos].ID == id {
+							newDefs = append(newDefs, pi.Definitions[defPos])
+							break
+						}
+					}
+				}
+				pi.Definitions = newDefs
+			} else {
+				restoredDroppingDefinitions := false
+				for _, def := range pi.AddingDefinitions {
+					found := false
+					for i := range pi.Definitions {
+						if def.ID == pi.Definitions[i].ID {
+							if i == len(pi.Definitions)-1 {
+								pi.Definitions = pi.Definitions[:i]
+								if !restoredDroppingDefinitions {
+									pi.Definitions = append(pi.Definitions, pi.DroppingDefinitions...)
+									restoredDroppingDefinitions = true
+								}
+							} else {
+								if !restoredDroppingDefinitions {
```

Comment:
It will reorg `pi.Definitions` like `[pi.Definitions[0], pi.Definitions[1], ..., pi.DroppingDefinitions, ... pi.Definitions[n]]`.

Why we need insert dropping definitions after first adding definitions?

Reply:
Since RANGE partitioning can only REORGANIZE consecutive partitions and HASH/KEY only ADD/COALESCE partitions (and ALTER TABLE t PARTITION BY/ALTER TABLE t REMOVE PARTITIONING) will affect all partitions, the first new partition will be removed and replaced by the DroppingDefinitions.
But that shows a bug in this code, that it only replaces the first 'new' partition, not the full range...
So I will remove this code and just rely on one implementation, i.e. reuse the LIST partition code above, to make maintenance easier.